pkg/blobcache/blobcache.go

```diff
@@ -0,0 +1,571 @@
+	_ types.ImageReference   = &blobCacheReference{}
+	_ types.ImageSource      = &blobCacheSource{}
+	_ types.ImageDestination = &blobCacheDestination{}
+)
+
+const (
+	compressedNote   = ".compressed"
+	decompressedNote = ".decompressed"
+)
+
+// BlobCache is an object which saves copies of blobs that are written to it while passing them
+// through to some real destination, and which can be queried directly in order to read them
+// back.
+type BlobCache interface {
+	types.ImageReference
+	// HasBlob checks if a blob that matches the passed-in digest (and
+	// size, if not -1), is present in the cache.
+	HasBlob(types.BlobInfo) (bool, int64, error)
+	// Directories returns the list of cache directories.
+	Directory() string
+	// ClearCache() clears the contents of the cache directories.  Note
+	// that this also clears content which was not placed there by this
+	// cache implementation.
+	ClearCache() error
+}
+
+type blobCacheReference struct {
+	reference types.ImageReference
+	// WARNING: The contents of this directory may be accessed concurrently,
+	// both within this process and by multiple different processes
+	directory string
+	compress  types.LayerCompression
+}
+
+type blobCacheSource struct {
+	reference *blobCacheReference
+	source    types.ImageSource
+	sys       types.SystemContext
+	// this mutex synchronizes the counters below
+	mu          sync.Mutex
+	cacheHits   int64
+	cacheMisses int64
+	cacheErrors int64
+}
+
+type blobCacheDestination struct {
+	reference   *blobCacheReference
+	destination types.ImageDestination
+}
+
+func makeFilename(blobSum digest.Digest, isConfig bool) string {
+	if isConfig {
+		return blobSum.String() + ".config"
+	}
+	return blobSum.String()
+}
+
+// CacheLookupReferenceFunc wraps a BlobCache into a
+// types.LookupReferenceFunc to allow for using a BlobCache during
+// image-copy operations.
+func CacheLookupReferenceFunc(directory string, compress types.LayerCompression) types.LookupReferenceFunc {
+	// NOTE: this prevents us from moving BlobCache around and generalizes
+	// the libimage API.
+	return func(ref types.ImageReference) (types.ImageReference, error) {
+		ref, err := NewBlobCache(ref, directory, compress)
+		if err != nil {
+			return nil, errors.Wrapf(err, "error using blobcache %q", directory)
+		}
+		return ref, nil
+	}
+}
+
+// NewBlobCache creates a new blob cache that wraps an image reference.  Any blobs which are
+// written to the destination image created from the resulting reference will also be stored
+// as-is to the specified directory or a temporary directory.  The cache directory's contents
+// can be cleared by calling the returned BlobCache()'s ClearCache() method.
+// The compress argument controls whether or not the cache will try to substitute a compressed
+// or different version of a blob when preparing the list of layers when reading an image.
+func NewBlobCache(ref types.ImageReference, directory string, compress types.LayerCompression) (BlobCache, error) {
```

Comment:
The cache is populated in `PutBlob()` (`ImageDestination`) and consulted in `GetBlob()` (`ImageSource`).  It could conceivably also have populated the cache on `GetBlob()`, but it didn't occur to me to do that at the time.
An `ImageReference` can be asked to create both `ImageSource`s and `ImageDestination`s.  I didn't see much benefit in having two functions which took an `ImageReference` argument and returned a different `ImageReference` when the implementation logic didn't require them to be different.  If an `ImageSource` was also writing to the cache, I might have thought differently.

Reply:
tl;dr — skip to the bottom

---

Personally I just found it hard to think about the properties… let me see if I get this right:

## On (new, pull):

Activated through `libimageOptions.DestinationLookupReferenceFunc`, i.e. a c/storage `storageDest` is cache-wrapped, the registry `dockerSrc` isn’t. The `compress` argument to `NewBlobCache` is ignored because the cache isn’t used on a source (and anyway, it is set to `PreserveOriginal` which means “don’t change `LayerInfosForCopy`”).

- The original manifest + `dockerSrc.LayerInfosForCopy` is used without modification
- `storageDest.TryReusingBlob`, _if_ the layer isn’t in c/storage already, checks for a cached blob; if so, it writes it to `storageDest`, and we don’t contact the registry. (This can AFAICS happen only if the cache is longer-lived than the c/storage backend.)
- `dockerSrc.GetBlob` just contacts the registry
- `storageDest.PutBlob` always stores the original layer blob to the cache, and if it is gzip-compressed, also stores an uncompressed variant to the cache.

I.e. on pull, we almost always record the compressed versions in cache (on the first pull of that layer), and usually the uncompressed version.

## On push:

Activated through `libimageOptions.SourceLookupReferenceFunc`, i.e. a c/storage `storageSrc` is cache-wrapped, the registry `dockerDest` isn’t. `compress` defaults to `types.Compress` unless the push invoker disables compression.

- `storageSrc.LayerInfosForCopy` starts with the `storageSrc` version, but can replace digests of uncompressed storage layers with digests of cached compressed versions.
- `dockerDest.TryReusingBlob` is not modified, but it can check the registry first for the compressed version per the update above, before falling back to BlobInfoCache. [The difference is that `TryReusingBlob` always queries the registry for the primary digest, but `BlobInfoCache`-known associations are, currently, only used if we _saw_ that blob in the same registry. So using a compressed-digest version for the physical query has a higher chance of avoiding an upload.]
- `storageSrc.GetBlob` first checks the cache, which helps both for a compressed artifact (no need to re-compress) and an uncompressed one (no need to traverse the graph driver’s fragmented storage and re-assemble tar-split; the uncompressed data is then compressed by the generic code).
- `dockerDest.PutBlob` just contacts the registry

I.e. on push, for any pre-existing layers we first check for an on-registry version, and if it isn’t there, we push them from the cache, not reading from `c/storage.Store.Diff`. Only new layers are compressed+written — and we _don’t_ update the cache with any newly created layers (uncompressed nor compressed).

## On commit:

(commit source = `buildah.containerImageSource`; reads from a prepared blob or from c/storage. commit destination = _usually_ a c/storage destination, but not quite _always_.)

Activated manually; _both_ the `containerSrc` source and `commitDest` destination is cache-wrapped. `compress` defaults to `types.PreserveOriginal` unless the commit invoker enables compression.

- `containerSrc.LayerInfosForCopy` starts with the `containerImageSource` version (= the newly-built manifest contents); it can replace digests of uncompressed storage layers with digests of cached compressed versions, but it doesn’t do so by default.
- `commitDest.TryReusingBlob`, _if_ the layer isn’t in the destination already, checks for a cached blob; if so, it writes it to `commitDest`, and we don’t use the `containerSrc` version. (This can AFAICS basically never happen for a c/storage commit destination, where the layer either exists in the same c/storage, or it is entirely new; we would read the cache only for other destination types.)
- `containerSrc.GetBlob` first checks the cache, which does nothing for a c/storage destination — either the layer already exists there and we don’t get to GetBlob, or it is genuinely new and not in the cache. For other commit destinations, this potentially helps both for a compressed artifact (no need to re-compress) and an uncompressed one (no need to traverse the graph driver’s fragmented storage and re-assemble tar-split; the generic code can compress).
- `commitDest.PutBlob`, _if_ the blob is not being compressed on the fly, stores the original layer blob to the cache (it is not gzip-compressed as coming from from `containerSrc`). If it _is_ being compressed on the fly, it is not stored in the cache.

I.e. for a c/storage commit destination, this has the net effect of storing the uncompressed version of any _new_ layers to the cache (and that in turn means that a future push won’t need a `Store.Diff` to read that). For non-c/storage commit destinations, this uses the cache similarly to a push, using it to optimize writes of previously-pulled layers (but not any commit-created layers).

So, _in principle_,
- on a commit to c/storage, we would only need to cache-wrap the destination (which would be usefully restricted, “BlobCache is only ever used to wrap c/image/storage”)
- direct commits to non-c/storage behave more like pushes (where we want to wrap the source); in _rare_ cases (direct commits without compression) we potentially benefit from caching the commit outcome, which happens in the `PutBlob` destination, but mostly we only need to cache-wrap the source.

---

So the pull/push cases are straightforward; I was originally primarily focusing on the commit case (the only direct use of `NewCache`), which is non-obvious, tends to do nothing for _most_ layers, but what it does do (avoid a second `.Diff` for pushing newly created layers) might actually be quite relevant for performance.

---

Conceptually I quite like the idea that “the cache is an implementation optimization of a c/storage user (both `c/storage/image` and `buildah.containerImageSrc`)”, where both `PutBlob` and `GetBlob` can update the cache. Going even further …

_Hypothetically_, in some far-future idealized world, the cache could be integrated into `c/image/storageImageDestination` so that its `PutBlob` doesn’t need to do any private copies just to compute digests, and into `buildah.containerImageSrc` so that it can write the uncompressed blobs for the new layers directly into the cache instead of making a private copy first. That might be a performance improvement for low-memory / IO-starved systems. And integration of _that_ kind would look so completely different from the current `ImageReference` wrapper (direct cache support in the two transports) that it’s really not worth worrying about — we would probably provide a “cache” object with some kind of “get blob” / “record blob” API. So that’s clearly out of scope and not worth worrying about short-term, apart from maybe defining that “cache” object now.

So, yes, I was thinking about this wrong — splitting the uses into “producer”/“consumer”, as I was suggesting originally, is not really the right approach. The cache would be inherently better if it were directly integrated, which is a completely different API.

---

So, on balance, let’s keep the one wrapping function here for now, and just document that
- it updates the cache on `PutBlob` and `PutManifest` (destinations), reads from it on `GetBlob` , `GetManifest` (sources) and `TryReusingBlob` (destinations) — that is enough for callers to get the effects they _need_,
- it might use the cache in more ways in the future — that gives us flexibility.

Separately, I think it still makes sense to introduce a `BlobCache` type, with a `WrapImageReference` method replacing this `NewBlobCache` function, but that’s a weak preference and not blocking I guess.